spi/transfer-spi/src/main/java/org/eclipse/dataspaceconnector/spi/transfer/command/CommandHandler.java

```diff
@@ -0,0 +1,18 @@
+package org.eclipse.dataspaceconnector.spi.transfer.command;
+
+import org.eclipse.dataspaceconnector.spi.ObjectFactory;
+
+/**
+ * CommandHandlers are ephemeral objects that receive a {@link Command} object and act on it.
+ * They must have an empty default constructor because they will get instantiated through reflection by the {@link ObjectFactory}. Also, they MUST
+ * declare all their fields as {@link org.eclipse.dataspaceconnector.spi.system.Inject}.
+ *
+ * @param <T> The concrete type of {@link Command}
+ * @see ObjectFactory
+ */
+public interface CommandHandler<T extends Command> {
```

Comment:
What is the motivation for keeping command data and logic separated (or as mentioned in the PR description "clear separation between command ("what") and handler ("how"))? Keeping both data and behaviour in the same command class seems like a good encapsulation practice to me. The current implementation reminds me of an [anemic domain model](https://martinfowler.com/bliki/AnemicDomainModel.html).

Reply:
The `command` is a "declaration of intent", it is a request of an action, it can be represented by a simple data object. 
Keeping the commands separated from the handlers in our case is fundamental if we want to make the command serializable, to permit the implementation of a persistent queue for example.

An anemic domain model is a model where the domain model objects (like `TransferProcess`) are treated like POJO and all the logic (like, changing state etc...) is delegated to other services, definitely not our case.

Reply:
> IMO this change deserves a sample or integration test showing the command functionality working E2E. Besides some additional documentation would be required for developers to understand that commands should be used instead of directly using `TransferProcessStore` for modifying TransferProcesses.

Yup, a sample demonstrating the command queue is definitely a great idea. Will create one. Also, documentation about this new feature is non-existent at this time, so that'll go on my list too :)

Reply:
`The command is a "declaration of intent", it is a request of an action, it can be represented by a simple data object.
Keeping the commands separated from the handlers in our case is fundamental if we want to make the command serializable, to permit the implementation of a persistent queue for example.`

Naming seems a bit off in this case... `Command` implies that there is behaviour in it (it is a behavioural design pattern in the end). Consider renaming to match the real purpose of each class.

Reply:
I think you are talking about the classic GoF command pattern, but this is not an application of that pattern, it's an excerpt from the CQRS pattern, which provides the existence of `commands` as data objects and `command handlers` as behaviour classes. I find it very clever and neat, I used it a lot in the past.
What names would you suggest for these two?

Reply:
Ok I see. `Command` and `CommandHandler` are just very broad, as a developer I would not know what exactly I can do with it just by looking at the code + javadoc. As mentioned above a sample and documentation on how this feature should be used will definitely help me understanding (as well as other EDC developers). I think it is key to make it clear that commands are required to perform operations on persistent objects (like TransferProcesses) to prevent concurrency issues. This might be enough and no class renaming is necessary then.

Reply:
I added the sample `samples/04.2-modify-transferprocess`. Does that help in understanding? Also, are you ok with the naming?

Reply:
It certainly does, thanks for the efforts @paullatzelsperger . Naming is fine but as `Command` and `CommandHandler` are such generic concepts I would move these classes out of the `transfer-spi` module and into the `core-spi`. WDYT?

Reply:
not a bad idea, as we might re-use them later in the contract negotiation manager.